File: src/main/cpp/io/SparkMaxMotorIO.cpp
#include "io/SparkMaxMotorIO.h"

using namespace rev::spark;

SparkMaxMotorIO::SparkMaxMotorIO(
    int canID, config::SparkMaxConfig config,
    const std::string& motorType, std::string motorName)
    : SparkMaxMotorIO(canID, std::move(config), motorType, std::move(motorName),
                      EncoderMode::PRIMARY, std::nullopt, std::nullopt) {}

SparkMaxMotorIO::SparkMaxMotorIO(
    int canID, config::SparkMaxConfig config,
    const std::string& motorType, std::string motorName,
    config::AbsoluteEncoderConfig absEncoderConfig)
    : SparkMaxMotorIO(canID, std::move(config), motorType, std::move(motorName),
                      EncoderMode::ABSOLUTE, std::make_optional(std::move(absEncoderConfig)),
                      std::nullopt) {}

SparkMaxMotorIO::SparkMaxMotorIO(
    int canID, config::SparkMaxConfig config,
    const std::string& motorType, std::string motorName,
    config::AlternateEncoderConfig altEncoderConfig)
    : SparkMaxMotorIO(canID, std::move(config), motorType, std::move(motorName),
                      EncoderMode::ALTERNATE, std::nullopt,
                      std::make_optional(std::move(altEncoderConfig))) {}

SparkMaxMotorIO::SparkMaxMotorIO(
    int canID, config::SparkMaxConfig config,
    const std::string& motorType, std::string motorName,
    EncoderMode encoderMode,
    std::optional<config::AbsoluteEncoderConfig> absEncoderConfig,
    std::optional<config::AlternateEncoderConfig> altEncoderConfig)
    : m_motor{canID, SparkMax::MotorType::kBrushless},
      m_canID{canID},
      m_config{std::move(config)},
      m_motorType{motorType},
      m_encoderMode{encoderMode},
      m_motorController{m_motor.GetClosedLoopController()} {
  m_motor.ClearFaults();

  if (absEncoderConfig.has_value()) {
    m_config.Apply(absEncoderConfig.value());
  }
  if (altEncoderConfig.has_value()) {
    m_config.Apply(altEncoderConfig.value());
  }

  m_motor.Configure(m_config, SparkBase::ResetMode::kNoResetSafeParameters,
                    SparkBase::PersistMode::kPersistParameters);
  m_motor.GetEncoder().SetPosition(0);
}

void SparkMaxMotorIO::SetMotorVoltage(double voltage) {
  m_motor.SetVoltage(units::voltage::volt_t{voltage});
}

void SparkMaxMotorIO::SetMotorPercentage(double percentage) {
  m_motor.Set(percentage);
}

void SparkMaxMotorIO::SetMotorRPM(double rpm) {
  m_motorController.SetReference(rpm, SparkBase::ControlType::kMAXMotionVelocityControl);
}

void SparkMaxMotorIO::SetMotorPosition(double setpoint, int slotID) {
  m_motorController.SetReference(
      setpoint, SparkBase::ControlType::kMAXMotionPositionControl,
      static_cast<ClosedLoopSlot>(slotID));
}

void SparkMaxMotorIO::UpdateInputs(MotorIOInputs& inputs) {
  inputs.motorConnected = (m_motor.GetDeviceId() == m_canID);
  inputs.motorVoltage = m_motor.GetAppliedOutput() * m_motor.GetBusVoltage();
  inputs.motorCurrent = m_motor.GetOutputCurrent();
  inputs.motorTemperature = m_motor.GetMotorTemperature();
  inputs.motorVelocity = m_motor.GetEncoder().GetVelocity();

  switch (m_encoderMode) {
    case EncoderMode::ABSOLUTE:
      inputs.motorPosition = m_motor.GetAbsoluteEncoder().GetPosition();
      break;
    case EncoderMode::ALTERNATE:
      inputs.motorPosition = m_motor.GetAlternateEncoder().GetPosition();
      break;
    case EncoderMode::PRIMARY:
    default:
      inputs.motorPosition = m_motor.GetEncoder().GetPosition();
      break;
  }
}

File: src/main/java/frc/robot/io/SparkMaxMotorIO.java
package frc.robot.io;

import com.revrobotics.PersistMode;
import com.revrobotics.ResetMode;
import com.revrobotics.spark.ClosedLoopSlot;
import com.revrobotics.spark.SparkBase.ControlType;
import com.revrobotics.spark.SparkClosedLoopController;
import com.revrobotics.spark.SparkLowLevel.MotorType;
import com.revrobotics.spark.SparkMax;
import com.revrobotics.spark.config.AbsoluteEncoderConfig;
import com.revrobotics.spark.config.AlternateEncoderConfig;
import com.revrobotics.spark.config.SparkMaxConfig;
import edu.wpi.first.units.measure.Voltage;
import java.util.Optional;

/**
 * REV Spark MAX implementation of {@link MotorIO}.
 *
 * <p>Velocity setpoints use RPM. Position setpoints use the configured encoder's position units.
 * Spark MAX does not support the torque-current control mode, so that capability lives in
 * {@link TorqueCurrentMotorIO} instead of this class.
 */
public class SparkMaxMotorIO implements MotorIO {
  /** Encoder source used for position telemetry. */
  protected enum EncoderMode {
    PRIMARY,
    ABSOLUTE,
    ALTERNATE
  }

  protected final SparkMax motor;
  protected final int canID;
  protected final SparkMaxConfig config;
  protected final String motorType;

  private final SparkClosedLoopController motorController;
  private final EncoderMode encoderMode;

  /** Creates a Spark MAX using the primary relative encoder. */
  public SparkMaxMotorIO(int canID, SparkMaxConfig config, String motorType, String motorName) {
    this(
        canID,
        config,
        motorType,
        motorName,
        EncoderMode.PRIMARY,
        Optional.empty(),
        Optional.empty());
  }

  /** Creates a Spark MAX using the absolute encoder. */
  public SparkMaxMotorIO(
      int canID,
      SparkMaxConfig config,
      String motorType,
      String motorName,
      AbsoluteEncoderConfig absEncoderConfig) {
    this(
        canID,
        config,
        motorType,
        motorName,
        EncoderMode.ABSOLUTE,
        Optional.of(absEncoderConfig),
        Optional.empty());
  }

  /** Creates a Spark MAX using the alternate encoder. */
  public SparkMaxMotorIO(
      int canID,
      SparkMaxConfig config,
      String motorType,
      String motorName,
      AlternateEncoderConfig altEncoderConfig) {
    this(
        canID,
        config,
        motorType,
        motorName,
        EncoderMode.ALTERNATE,
        Optional.empty(),
        Optional.of(altEncoderConfig));
  }

  /** Full constructor for selecting the encoder source and optional encoder config. */
  public SparkMaxMotorIO(
      int canID,
      SparkMaxConfig config,
      String motorType,
      String motorName,
      EncoderMode encoderMode,
      Optional<AbsoluteEncoderConfig> absEncoderConfig,
      Optional<AlternateEncoderConfig> altEncoderConfig) {
    this.canID = canID;
    this.config = config;
    this.motorType = motorType;
    this.encoderMode = encoderMode;

    motor = new SparkMax(this.canID, MotorType.kBrushless);
    motorController = motor.getClosedLoopController();

    motor.clearFaults();

    applyEncoderConfigs(absEncoderConfig, altEncoderConfig);

    motor.configure(this.config, ResetMode.kNoResetSafeParameters, PersistMode.kPersistParameters);
    motor.getEncoder().setPosition(0);
  }

  @Override
  public void setMotorVoltage(Voltage voltage) {
    motor.setVoltage(voltage);
  }

  @Override
  public void setMotorPercentage(double percentage) {
    motor.set(percentage);
  }

  @Override
  public void setMotorRPM(double rpm) {
    motorController.setSetpoint(rpm, ControlType.kMAXMotionVelocityControl);
  }

  @Override
  public void setMotorPosition(double setpoint, int slotID) {
    motorController.setSetpoint(
        setpoint, ControlType.kMAXMotionPositionControl, ClosedLoopSlot.values()[slotID]);
  }

  @Override
  public void updateInputs(MotorIOInputs inputs) {
    inputs.setMotorConnected(motor.getDeviceId() == canID);
    inputs.setMotorVoltage(motor.getAppliedOutput() * motor.getBusVoltage());
    inputs.setMotorCurrent(motor.getOutputCurrent());
    inputs.setMotorTemperature(motor.getMotorTemperature());
    inputs.setMotorVelocity(motor.getEncoder().getVelocity());
    inputs.setMotorPosition(readSelectedEncoderPosition());
  }

  private void applyEncoderConfigs(
      Optional<AbsoluteEncoderConfig> absEncoderConfig,
      Optional<AlternateEncoderConfig> altEncoderConfig) {
    absEncoderConfig.ifPresent(this.config::apply);
    altEncoderConfig.ifPresent(this.config::apply);
  }

  private double readSelectedEncoderPosition() {
    return switch (encoderMode) {
      case ABSOLUTE -> motor.getAbsoluteEncoder().getPosition();
      case ALTERNATE -> motor.getAlternateEncoder().getPosition();
      case PRIMARY -> motor.getEncoder().getPosition();
    };
  }
}
